src/dimension.c

```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid interval: must be an interval or integer type")));
+	}
+}
+
+static void
+dimension_add_not_null_on_column(Oid table_relid, char *colname)
+{
+	AlterTableCmd cmd = {
+		.type = T_AlterTableCmd,
+		.subtype = AT_SetNotNull,
+		.name = colname,
+		.missing_ok = false,
+	};
+
+	elog(NOTICE, "Adding NOT NULL constraint to time column %s (NULL time values not allowed)", colname);
+
+	AlterTableInternal(table_relid, list_make1(&cmd), false);
+}
+
+static void
+dimension_update(FunctionCallInfo fcinfo,
+				 Oid table_relid,
+				 Name dimname,
+				 DimensionType dimtype,
+				 Datum *interval,
+				 int16 *num_slices)
+{
+	Cache	   *hcache = hypertable_cache_pin();
+	Hypertable *ht;
+	Dimension  *dim;
+
+	if (dimtype == DIMENSION_TYPE_ANY)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid dimension type")));
+
+	ht = hypertable_cache_get_entry(hcache, table_relid);
+
+	if (NULL == ht)
+		ereport(ERROR,
+				(errcode(ERRCODE_IO_HYPERTABLE_NOT_EXIST),
+				 errmsg("table \"%s\" is not a hypertable",
+						get_rel_name(table_relid))));
+
+	if (NULL == dimname)
+	{
+		if (hyperspace_get_num_dimensions_by_type(ht->space, dimtype) > 1)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("hypertable \"%s\" has multiple %s dimensions",
+							get_rel_name(table_relid),
+							dimtype == DIMENSION_TYPE_OPEN ? "time" : "space"),
+					 errhint("An explicit dimension name needs to be specified")));
+
+		dim = hyperspace_get_dimension(ht->space, dimtype, 0);
+	}
+	else
+		dim = hyperspace_get_dimension_by_name(ht->space, dimtype, NameStr(*dimname));
+
+	if (NULL == dim)
+		ereport(ERROR,
+				(errcode(ERRCODE_IO_DIMENSION_NOT_EXIST),
+				 errmsg("hypertable \"%s\" does not have a matching dimension",
+						get_rel_name(table_relid))));
+
+	if (NULL != interval)
+	{
+		Oid			intervaltype = get_fn_expr_argtype(fcinfo->flinfo, 1);
+
+		dim->fd.interval_length = dimension_interval_to_internal(dim->fd.column_type,
+																 intervaltype,
+																 *interval,
+																 true);
+	}
+
+	if (NULL != num_slices)
+		dim->fd.num_slices = *num_slices;
+
+	dimension_scan_update(dim->fd.id, dimension_tuple_update, dim, RowExclusiveLock);
+
+	cache_release(hcache);
+}
+
+TS_FUNCTION_INFO_V1(dimension_set_num_slices);
+
+Datum
+dimension_set_num_slices(PG_FUNCTION_ARGS)
+{
+	Oid			table_relid = PG_GETARG_OID(0);
+	int32		num_slices_arg = PG_ARGISNULL(1) ? -1 : PG_GETARG_INT32(1);
+	Name		colname = PG_ARGISNULL(2) ? NULL : PG_GETARG_NAME(2);
+	int16		num_slices;
+
+	hypertable_permissions_check(table_relid, GetUserId());
+
+	if (PG_ARGISNULL(1) || num_slices_arg < 1 || num_slices_arg > INT16_MAX)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid number of partitions")));
+
+	num_slices = num_slices_arg & 0xffff;
```

Comment:
So, the in-argument for the function is an `INTEGER` (32 bits) although num_slices is an `SMALLINT` in our catalog. We use an `INTEGER` argument avoid a cast when calling the API function. The value of the `num_slices_arg` `int32` is verified to be within the bounds of an `int16`, then it can be safely converted to an `int16`. Without this stuff, the compiler would complain as `dimension_update()` and the catalog table takes an `int16`. I could do a C-style cast if you prefer that.

Reply:
Yes I think a c-style cast is more clear.